src/experiment.rs

```diff
@@ -0,0 +1,124 @@
+    /// The configuration variables.
+    config: Config,
+    /// The benchmarks to run.
+    benchmarks: Vec<&'a Benchmark<'a>>,
+}
+
+impl<'a> Experiment<'a> {
+    // Private: experiments should always be created through the ExperimentBuilder.
+    fn new(config: Config, benchmarks: Vec<&'a Benchmark>) -> Self {
+        Experiment { config, benchmarks }
+    }
+
+    /// Run the experiment. If experiment completes successfully, return a String
+    /// which represents the path of the results file; otherwise, return a `K2Error`.
+    pub fn run(self) -> Result<String, K2Error> {
+        unimplemented!("run");
+    }
+}
+
+/// A builder used to construct an `Experiment`.
+///
+/// This populates a `Config` struct with values, and collects the benchmarks
+/// to run.
+pub struct ExperimentBuilder<'a> {
+    config: Config,
+    benchmarks: Vec<&'a Benchmark<'a>>,
+}
+
+impl<'a> ExperimentBuilder<'a> {
+    pub fn new() -> Self {
+        ExperimentBuilder {
+            config: Config::new(),
+            benchmarks: Default::default(),
+        }
+    }
+
+    pub fn result_path(mut self, result_path: &str) -> Self {
+        self.config.result_path = result_path.to_string();
+        self
+    }
+
+    pub fn quick(mut self, quick: bool) -> Self {
+        self.config.quick = quick;
+        self
+    }
+
+    pub fn dry_run(mut self, dry_run: bool) -> Self {
+        self.config.dry_run = dry_run;
+        self
+    }
+
+    pub fn reboot(mut self, reboot: bool) -> Self {
+        self.config.reboot = reboot;
+        self
+    }
+
+    pub fn mail_to(mut self, mail_to: Vec<String>) -> Self {
+        self.config.mail_to = mail_to;
+        self
+    }
+
+    pub fn in_proc_iters(mut self, in_proc_iters: usize) -> Self {
+        self.config.in_proc_iters = in_proc_iters;
+        self
+    }
+
+    pub fn pexecs(mut self, pexecs: usize) -> Self {
+        self.config.pexecs = pexecs;
+        self
+    }
+
+    pub fn stack_lim(mut self, stack_lim: Limit) -> Self {
+        self.config.stack_lim = Some(stack_lim);
+        self
+    }
+
+    pub fn heap_lim(mut self, heap_lim: Limit) -> Self {
+        self.config.heap_lim = Some(heap_lim);
+        self
+    }
+
+    pub fn temp_read_pause(mut self, temp_read_pause: Duration) -> Self {
+        self.config.temp_read_pause = temp_read_pause;
+        self
+    }
+
+    pub fn enable_pinning(mut self, enable_pinning: bool) -> Self {
+        self.config.enable_pinning = enable_pinning;
+        self
+    }
+
+    pub fn wallclock(mut self, wallclock: bool) -> Self {
+        self.config.wallclock = wallclock;
+        self
+    }
+
+    pub fn core_cycles(mut self, core_cycles: bool) -> Self {
+        self.config.core_cycles = core_cycles;
+        self
+    }
+
+    pub fn perf(mut self, perf: bool) -> Self {
+        self.config.perf = perf;
+        self
+    }
+
+    /// Add `bench` to the list of benchmarks to run.
+    pub fn add_benchmark(mut self, bench: &'a Benchmark) -> Self {
```

Comment:
Should this be just `benchmark()` to be in line with the other options?

Also in `Command` you have `.arg().arg()` instead of `add_arg().add_arg()`:
https://doc.rust-lang.org/std/process/struct.Command.html

Reply:
You're right, `benchmark()` is much better.